compiler/quilt/asa/img.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2018, Quilt Data Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#    http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Convenience functions for displaying images in Jupyter notebooks.
+
+`pip install quilt[img]`
+
+Or, in development:
+`pip install -e ./[img]`
+"""
+
+from math import ceil, floor, sqrt
+from six import string_types
+
+import matplotlib.image as mpimg
+import matplotlib.pyplot as plt
+
+from quilt.tools.const import ELLIPSIS
+from quilt.nodes import DataNode, GroupNode
+from quilt.tools.build import splitext_no_dot
+
+def plot(figsize=None, formats=None, limit=100, titlelen=10, **kwargs):
+    """Display an image [in a Jupyter Notebook] from a Quilt fragment path.
+    Intended for use with `%matplotlib inline`.
+
+    Convenience method that loops over supblots that call
+    `plt.imshow(image.imread(FRAG_PATH))`.
+
+    Keyword arguments
+    * figsize=None # None means auto, else provide (HEIGHT_INCHES, WIDTH_INCHES)
+    * formats=None # List of extensions as strings ['jpg', 'png', ...]
+    * limit=100 # maximum number of images to display
+    * titlelen=10 # max number of characters in subplot title
+    * **kwargs - all remaining kwargs are passed to plt.subplots;
+      see https://matplotlib.org/api/_as_gen/matplotlib.pyplot.subplots.html
+    """
+    # pylint: disable=protected-access
+    def _plot(node, paths):
+        lower_formats = set((x.lower() for x in formats)) if formats !=None else None
+        def node_filter(_, frag, meta):
+            filepath = meta.get('_system', {}).get('filepath', None)
+            # don't try to read DataFrames as images
+            if isinstance(frag, string_types) and filepath:
+                _, ext = splitext_no_dot(filepath)
+                if formats is None or ext.lower() in lower_formats:
```

Comment:
Just for consistency, I'd use `lower_formats is None` instead of `formats is None`